lib/src/package_config_impl.dart

```diff
@@ -0,0 +1,181 @@
+
+  static int _validateVersion(int version) {
+    RangeError.checkValueInInterval(version, 1, PackageConfig.maxVersion, "version");
+    return version;
+  }
+
+  static Map<String, Package> _validatePackages(Iterable<Package> packages) {
+    Map<String, Package> result = {};
+    for (var package in packages) {
+      if (package is! _SimplePackage) {
+        // _SimplePackage validates these properties.
+        try {
+          _validatePackageData(package.name, package.root,
+              package.packageUriRoot, package.languageVersion);
+        } catch (e) {
+          throw ArgumentError.value(
+              packages, "packages", "Package ${package.name}: ${e.message}");
+        }
+      }
+      var name = package.name;
+      if (result.containsKey(name)) {
+        throw ArgumentError.value(name, "packages", "Duplicate package name");
+      }
+      result[name] = package;
+    }
+
+    // Check that no root URI is a prefix of another.
+    if (result.length > 1) {
+      // Uris cache their toString, so this is not as bad as it looks.
+      var rootUris = [...result.values]
+        ..sort((a, b) => a.root.toString().compareTo(b.root.toString()));
+      var prev = rootUris[0];
+      var prevRoot = prev.root.toString();
+      for (int i = 1; i < rootUris.length; i++) {
+        var next = rootUris[i];
+        var nextRoot = next.root.toString();
+        // If one string is a prefix of another,
+        // the former sorts just before the latter.
+        if (nextRoot.startsWith(prevRoot)) {
+          throw ArgumentError.value(
+              packages,
+              "packages",
+              "Package ${next.name} root overlaps "
+                  "package ${prev.name} root.\n"
+                  "${prev.name} root: $prevRoot\n"
+                  "${next.name} root: $nextRoot\n");
+        }
+        prev = next;
+      }
+    }
+    return result;
+  }
+
+  Iterable<Package> get packages => _packages.values;
+
+  Package /*?*/ operator [](String packageName) => _packages[packageName];
+
+  /// Provides the associated package for a specific [file] (or directory).
+  ///
+  /// Returns a [Package] which contains the [file]'s path.
+  /// That is, the [Package.rootUri] directory is a parent directory
+  /// of the [file]'s location.
+  /// Returns `null` if the file does not belong to any package.
+  Package /*?*/ packageOf(Uri file) {
+    String path = file.toString();
+    for (var package in _packages.values) {
+      var rootPath = package.root.toString();
+      if (path.startsWith(rootPath)) return package;
+    }
+    return null;
+  }
+
+  Uri /*?*/ resolve(Uri packageUri) {
+    String packageName = checkValidPackageUri(packageUri);
+    return _packages[packageName]?.packageUriRoot?.resolveUri(
+        Uri(path: packageUri.path.substring(packageName.length + 1)));
+  }
+
+  Uri /*?*/ toPackageUri(Uri nonPackageUri) {
+    if (nonPackageUri.isScheme("package")) {
+      throw ArgumentError.value(
```

Comment:
How about returning it directly so we can avoid code like ` uri.isScheme('package') ? uri : packageConfig.toPackageUri(uri)`

Reply:
I'm a little worried about allowing that.

The function is defined as returning a package URI which resolves, using `PackageConfig.resolve`, to the input URI.

If the input is a package URI then that is no longer true.

I think I prefer to disallow this until I see a concrete need for the more permissive behavior.